### ISSUE TITLE
Silent fail: no visible stacktrace upon an exception

### ISSUE THREAD

**Issue by timur87** [clojure-emacs/cider#1404]
Hi, 

The exception occurs in a seperate thread created by a (future). Clojure version is 1.7.0. :pre :post conditions are also now shown.

No exception is shown on CIDER REPL or on nREPL server when execution fails with an exception: 

``` clojure
(defn- add-state
  [m]
  (debug "Test")
  (throw (Exception.)))
```

> 15-Nov-09 18:01:27 dyn125 DEBUG [test-prj] - Test

Some project details:
; CIDER 0.10.0snapshot (package: 20151106.2226) (Java 1.8.0_45, Clojure 1.7.0, nREPL 0.2.10 (12))

Any ideas how to fix this problem?

**Comment by bbatsov** [comment]
Can you share with us your `*nrepl-messages*` log?

**Comment by timur87** [comment]
Here is a simplified version of the last lines, hopefully it's so okay. There are huge maps in place of huge maps in places where I wrote huge map and some string in place of some string. After model addition the exception has been thrown with (throw (Exception.))

<addr>
(<- 
  id  "8"
  out  "15-Nov-09 22:29:44 linux-dv1t DEBUG [execution-environment-integrator] - Error sequence will be executed!! (huge map)"
  session  "a2411d18-f78b-4f62-8e36-350df1c07768"
)
(<- 
  id  "8"
  out  "\n"
  session  "a2411d18-f78b-4f62-8e36-350df1c07768"
)
(<- 
  id  "8"
  out  "15-Nov-09 22:29:44 linux-dv1t ERROR [execution-environment-integrator] - Error sequence will be executed!! (str)"
  session  "a2411d18-f78b-4f62-8e36-350df1c07768"
)
(<- 
  id  "8"
  out  "\n"
  session  "a2411d18-f78b-4f62-8e36-350df1c07768"
)
(<- 
  id  "8"
  out  "15-Nov-09 22:29:44 linux-dv1t DEBUG [execution-environment-integrator] - Adding final model state (huge map)"
  session  "a2411d18-f78b-4f62-8e36-350df1c07768"
)
(<- 
  id  "8"
  out  "\n"
  session  "a2411d18-f78b-4f62-8e36-350df1c07768"
)
</addr>

**Comment by bbatsov** [comment]
Yeah, I can reproduce this problem as well. Seems to that this is some nREPL bug/limitation. @cemerick any idea what's going on here?

**Comment by cemerick** [comment]
I can't load the code as provided; `debug` isn't defined. Changing it to `println` yields code I can load, but the exception is reported as expected.

**Comment by bbatsov** [comment]
Are you sure? Even an a regular lein repl I get this the first time I run the code:

```
user=> (future (do
  #_=>           (println "Test 1")
  #_=>           (throw (Exception.))
  #_=>           (println "Test 2")))
Test 1
#object[clojure.core$future_call$reify__6736 0x15dea85b {:status :failed, :val #error {
 :cause nil
 :via
 [{:type java.util.concurrent.ExecutionException
   :message "java.lang.Exception"
   :at [java.util.concurrent.FutureTask report "FutureTask.java" 122]}
  {:type java.lang.Exception
   :message nil
   :at [user$eval7994$fn__7995 invoke "form-init5463094278957846993.clj" 2]}]
 :trace
 [[user$eval7994$fn__7995 invoke "form-init5463094278957846993.clj" 2]
  [clojure.core$binding_conveyor_fn$fn__4444 invoke "core.clj" 1916]
  [clojure.lang.AFn call "AFn.java" 18]
  [java.util.concurrent.FutureTask run "FutureTask.java" 266]
  [java.util.concurrent.ThreadPoolExecutor runWorker "ThreadPoolExecutor.java" 1142]
  [java.util.concurrent.ThreadPoolExecutor$Worker run "ThreadPoolExecutor.java" 617]
  [java.lang.Thread run "Thread.java" 745]]}}]
```

If I try to execute it again it changes to this:

```
user=> (future (do
  #_=>           (println "Test 1")
  #_=>           (throw (Exception.))
  #_=>           (println "Test 2")))
Test 1
#object[clojure.core$future_call$reify__6736 0x69cf66e3 {:status :pending, :val nil}]
```

@Malabarba Btw, the first eval result looks horrible when displayed inline. :-)

**Comment by Malabarba** [comment]
I get the same behaviours here.

>  Btw, the first eval result looks horrible when displayed inline. :-)

What do you mean?

**Comment by bbatsov** [comment]
> What do you mean?

I mean that this is one huge result that didn't look very pretty in my Clojure buffer. If you eval this interactively you'll see what I mean.

**Comment by bbatsov** [comment]
I'm totally confused by the problem, so I'd love to hear some feedback/ideas. I still think that something's goes wrong with nREPL, but @cemerick seems to believe this is not the case. I also have absolutely now idea why re-evaluating the same `future` code twice produces different results. 

P.S. Dereferencing the future provides the expected results.

**Comment by expez** [comment]
#### CIDER repl: 

```
test.core> (future (do
                     (println "Test 1")
                     (throw (Exception.))
                     (println "Test 2")))
Test 1
#<Future@79c2353f: Exception   test.core/eval37719/fn--37720 (form-init8274982211765589402.clj:2)
```


Along with a popup stacktrace buffer

#### lein repl

```
user=> (future (do (println "Test 1") (throw (Exception.)) (println "Test 2")))
Test 1
#object[clojure.core$future_call$reify__6736 0x242640e5 {:status :pending, :val nil}]
```

I don't get why the stacktrace appears in CIDER (as I'm not dereferencing the future) but I'm not seeing anything else weird like non-deterministic results.

I even checked with 1.6 to see if that explained why @bbatsov was seeing something different, but that wasn't it.

**Comment by timur87** [comment]
@expez I don't think that the expected behavior of running this code is an object. Here, we have an uncatched exception, but it's not shown in a CIDER error frame If one wants to see the right behavior, he can execute the same expression w/o future. 

```clojure
(do
                     (println "Test 1")
                     (throw (Exception.))
                     (println "Test 2")))

```

Or am I mistaken by expecting to see the exception happened in a seperate thread?


Apart from that, having 2 different results for the consequent executions of the very same code is a proof of something is going wrong, either in my setting or with Cider. Moreover, if you do this without the future, you get an CIDER error in a separate frame and uncatched exception is shown to you with appropriate stack trace. So we have 3 different results from CIDER / nREPL for the very same expression. 

In case of a Ring applicaiton, one doesn't get to see the exception occuring in a handler which is quite contra-productive. I again post my results, that I have recieved two mins ago, if you believe it's not related to CIDER but nREPL, I will delegate this (bug) to there.

To reproduce the results:

```clojure
; CIDER 0.10.0snapshot (package: 20151125.1430) (Java 1.8.0_45, Clojure 1.7.0, nREPL 0.2.12)
user> (future (do
                     (println "Test 1")
                     (throw (Exception.))
                     (println "Test 2")))
Test 1
#object[clojure.core$future_call$reify__6736 0x2b013dd0 {:status :failed, :val #error {
 :cause nil
 :via
 [{:type java.util.concurrent.ExecutionException
   :message "java.lang.Exception"
   :at [java.util.concurrent.FutureTask report "FutureTask.java" 122]}
  {:type java.lang.Exception
   :message nil
   :at [user$eval13914$fn__13915 invoke "form-init2622773537481546772.clj" 3]}]
 :trace
 [[user$eval13914$fn__13915 invoke "form-init2622773537481546772.clj" 3]
  [clojure.core$binding_conveyor_fn$fn__4444 invoke "core.clj" 1916]
  [clojure.lang.AFn call "AFn.java" 18]
  [java.util.concurrent.FutureTask run "FutureTask.java" 266]
  [java.util.concurrent.ThreadPoolExecutor runWorker "ThreadPoolExecutor.java" 1142]
  [java.util.concurrent.ThreadPoolExecutor$Worker run "ThreadPoolExecutor.java" 617]
  [java.lang.Thread run "Thread.java" 745]]}}]
user> (future (do
                     (println "Test 1")
                     (throw (Exception.))
                     (println "Test 2")))
Test 1
#object[clojure.core$future_call$reify__6736 0xbc8bf2 {:status :pending, :val nil}]
user> 

(do
                     (println "Test 1")
                     (throw (Exception.))
                     (println "Test 2"))
Test 1
Exception   user/eval13924 (form-init2622773537481546772.clj:32)
user> 

;; lein repl result

user=> (future (do
  #_=>                      (println "Test 1")
  #_=>                      (throw (Exception.))
  #_=>                      (println "Test 2")))
Test 1
#object[clojure.core$future_call$reify__6736 0x28728772 {:status :pending, :val nil}]
user=> 

```

**Comment by expez** [comment]
> I don't think that the expected behavior of running this code is an object. Here, we have an uncatched exception, but it's not shown in a CIDER error frame If one wants to see the right behavior, he can execute the same expression w/o future.

The result of creating a future is indeed an 'object'.  To get the value out of a future, you have to dereference this 'object', as you know, using the `deref` function or the `@` reader macro.  If the future failed to produce a value, this dereferencing will re-throw the exception in the current thread.

> Or am I mistaken by expecting to see an exception happened in a seperate thread?

Yeah, you are mistaken.  This caught me by surprise the first time I saw it too (I had threads dying but got no feedback about a crash :P), but it kind of makes sense when you think about it.

The only thing I'm confused about is the varying behavior.

When I evaluate

```
(future (do
                     (println "Test 1")
                     (throw (Exception.))
                     (println "Test 2")))
```

The result is mostly: 
```
#<Future@1671daf0: :pending>
Test 1
```

But the first couple of times I evaluated this I saw an `ExcecutionException` stacktrace pop up.

The former is what I expect, and I can't explain the latter.  I'm inclined to just ignore this quirky behavior that sometimes occurs, as I've no idea how to fix it and it might take considerable work to get to the bottom of it.  Thoughts, @bbatsov?

**Comment by bbatsov** [comment]
I still think something's wrong with nREPL, as nothing else could account for the different results.

**Comment by cemerick** [comment]
Leaving aside CIDER's UI behaviour for a moment, replying to https://github.com/clojure-emacs/cider/issues/1404#issuecomment-156690096, seeing different printed results of the expression

```clojure
(future (do
          (println "Test 1")
          (throw (Exception.))
          (println "Test 2")))
```

is perfectly reasonable. The code that throws the exception is running in a future, and so if it's still `:pending` (as the second evaluation seems to be doing for many people), that just means that the exception hasn't been thrown by the time the REPL printed the representation of the future.

As to why CIDER isn't getting error information back on the still-pending future, I think perhaps there's either a problem with how nREPL sessions are being propagated (they're essentially just the dynamic environment + a few other bits and bobs, and so _should_ be carried along into futures via "binding conveyance", in the Clojure nomenclature), or some middleware used by the tooling to signal/handle errors is misbehaving or is sensitive to binding-conveying forms (agent sends have the same semantics, in case someone wants to try to replicate the behaviour using them).

**Comment by bhagany** [comment]
Also fixed in #387
